src/physics/base/ParticleDefContainer.cu

```diff
@@ -0,0 +1,99 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file ParticleDefContainer.cu
+//---------------------------------------------------------------------------//
+#include "ParticleDefContainer.hh"
+
+#include <thrust/device_vector.h>
+#include "base/Macros.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+struct ParticleDefContainerPimpl
+{
+    thrust::device_vector<ParticleDef> defs;
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * \brief Construct with a vector of particle definitions
+ */
+ParticleDefContainer::ParticleDefContainer(const Params& params)
+{
+    std::vector<ParticleDef> host_defs;
+    host_defs.reserve(params.defs.size());
+
+    for (const auto& md_data : params.defs)
+    {
+        // Add host metadata
+        ParticleDefId id(name_to_id_.size());
+        bool          inserted;
+        std::tie(std::ignore, inserted)
+            = name_to_id_.insert({md_data.first.name, id});
+        CHECK(inserted);
+        std::tie(std::ignore, inserted)
+            = pdg_to_id_.insert({md_data.first.pdg_code, id});
+        CHECK(inserted);
+
+        // Add device data to be copied
+        host_defs.emplace_back(md_data.second);
+    }
+
+    // Allocate and copy data to device
+    data_       = std::make_unique<ParticleDefContainerPimpl>();
+    data_->defs = host_defs;
+
+    ENSURE(name_to_id_.size() == params.defs.size());
+    ENSURE(pdg_to_id_.size() == params.defs.size());
+    ENSURE(data_->defs.size() == params.defs.size());
+}
+
+//---------------------------------------------------------------------------//
+// Default constructor/destructor/move
+ParticleDefContainer::ParticleDefContainer()                       = default;
```

Comment:
Wouldn't moving those in the class declaration improve in-lining?

Reply:
This class is only going to be constructed a handful of times at most (and probably only once). But the real reason to declare them non-inline is to hide the cuda calls so that the `.hh` can be included downstream.

Reply:
What cuda calls? (They are set to "default")

Reply:
Resolved (this is because of the pimpl).

Reply:
I was surprised as you are that the default constructor of `thrust::device_vector` calls CUDA -- turns out it calls `thrust::for_each` kernel since it's calling an internal `vec.assign(count)` even though the value of `count` is zero.